scipy/stats/_qmc.py

```diff
@@ -1,33 +1,52 @@
 """Quasi-Monte Carlo engines and helpers."""
-import os
+from __future__ import annotations
+
 import copy
-import numbers
-from abc import ABC, abstractmethod
 import math
+import numbers
+import os
 import warnings
+from abc import ABC, abstractmethod
+from typing import (
+    Optional,
+    Union,
+    TYPE_CHECKING
+)
 
 import numpy as np
 
-import scipy.stats as stats
-from scipy.stats._sobol import (
-    initialize_v, _cscramble, _fill_p_cumulative, _draw, _fast_forward,
-    _categorize, initialize_direction_numbers, _MAXDIM, _MAXBIT
-)
+if TYPE_CHECKING:
+    import numpy.typing as npt
+
 from scipy.stats._qmc_cy import (
     _cy_wrapper_centered_discrepancy,
     _cy_wrapper_wrap_around_discrepancy,
     _cy_wrapper_mixture_discrepancy,
     _cy_wrapper_l2_star_discrepancy,
     _cy_wrapper_update_discrepancy
 )
+from scipy.stats._sobol import (
+    initialize_v, _cscramble, _fill_p_cumulative, _draw, _fast_forward,
+    _categorize, initialize_direction_numbers, _MAXDIM, _MAXBIT
+)
+
+import scipy.stats as stats
+
 
 __all__ = ['scale', 'discrepancy', 'update_discrepancy',
            'QMCEngine', 'Sobol', 'Halton', 'LatinHypercube',
            'MultinomialQMC', 'MultivariateNormalQMC']
 
+IntNumber = Union[int, np.integer]
+DecimalNumber = Union[float, np.floating]
```

Comment:
These should be used as before through the file. Here is some background https://github.com/scipy/scipy/pull/10844#pullrequestreview-559980027

Reply:
The issue is, with the old types (in `_qmc.pyi`) we have got a lot of types errors when putting them into inline types : 
```Shell
scipy/stats/_qmc.py:400: error: Argument 1 to "ones" has incompatible type "Union[int, number[Any]]"; expected "Union[int, Sequence[int]]"  [arg-type]
scipy/stats/_qmc.py:435: error: Slice index must be an integer or None  [misc]
scipy/stats/_qmc.py:440: error: Slice index must be an integer or None  [misc]
scipy/stats/_qmc.py:492: error: Argument 1 to "zeros" has incompatible type "Union[int, integer[Any]]"; expected "Union[int, Sequence[int]]"  [arg-type]
scipy/stats/_qmc.py:773: error: Incompatible types in assignment (expression has type "Union[int, number[Any]]", variable has type "int")  [assignment]
scipy/stats/_qmc.py:774: error: Argument 1 to "reshape" of "_ArrayOrScalarCommon" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:774: error: Argument 2 to "reshape" of "_ArrayOrScalarCommon" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:873: error: Argument 1 to "range" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:878: error: Incompatible types in assignment (expression has type "Union[int, number[Any]]", variable has type "int")  [assignment]
scipy/stats/_qmc.py:1028: error: Argument 1 to "zeros" has incompatible type "Tuple[Union[int, integer[Any]], Any]"; expected "Union[int, Sequence[int]]"  [arg-type]
scipy/stats/_qmc.py:1032: error: Argument 1 to "zeros" has incompatible type "Union[int, integer[Any]]"; expected "Union[int, Sequence[int]]"  [arg-type]
scipy/stats/_qmc.py:1071: error: Argument 1 to "empty" has incompatible type "Tuple[Union[int, integer[Any]], Union[int, integer[Any]]]"; expected "Union[int, Sequence[int]]"  [arg-type]
scipy/stats/_qmc.py:1075: error: Unsupported operand types for & ("int" and "number[Any]")  [operator]
scipy/stats/_qmc.py:1075: error: Unsupported operand types for & ("integer[Any]" and "number[Any]")  [operator]
scipy/stats/_qmc.py:1075: note: Both left and right operands are unions
scipy/stats/_qmc.py:1084: error: Slice index must be an integer or None  [misc]
scipy/stats/_qmc.py:1089: error: Incompatible types in assignment (expression has type "Union[int, number[Any]]", variable has type "int")  [assignment]
scipy/stats/_qmc.py:1112: error: Unsupported left operand type for & ("number[Any]")  [operator]
scipy/stats/_qmc.py:1112: note: Both left and right operands are unions
scipy/stats/_qmc.py:1120: error: Argument 1 to "random" of "Sobol" has incompatible type "Union[Any, number[Any]]"; expected "Union[int, integer[Any]]"  [arg-type]
scipy/stats/_qmc.py:1303: error: unused 'type: ignore' comment
scipy/stats/_qmc.py:1304: error: Module has no attribute "norm"  [attr-defined]
scipy/stats/_qmc.py:1313: error: Argument 1 to "reshape" of "_ArrayOrScalarCommon" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:1315: error: Slice index must be an integer or None  [misc]
```

Reply:
@rgommers apparently mypy complains about `_IntegerType = Union[int, np.integer]` and the same for float. Is it fine then to use plain int?

Reply:
That's odd, it shouldn't matter if the annotation is inline or in a stub file. Just to make sure: this is with the most recent NumPy version?

```

scipy/stats/_qmc.py:492: error: Argument 1 to "zeros" has incompatible type "Union[int, integer[Any]]"; expected "Union[int, Sequence[int]]"  [arg-type]
```

The current definition of `_ShapeLike` is `Union[int, Sequence[int]]`, so that indeed doesn't accept numpy integer types. That may be an oversight - @BvB93 can you comment?

Reply:
That definitely seems to be an oversight, yes. At a quick glance `int` either has to be replaced with the `typing.SupportsIndex` protocol or, alternatively, a union of `int` and `np.integer`.

Reply:
FYI, a PR addressing the `_ShapeLike` issue is currently up at https://github.com/numpy/numpy/pull/18778.

Reply:
Thanks Bas!

Okay, so we want to keep `_IntegerType` the same here. It works with numpy ~master~ main now. It's just a little annoying that CI will fail. Or at least it should - it's not failing now, and we should figure out why cause apparently it's broken. 

We can then switch `.github/workflows/linux.yml` to use numpy nightlies to test, because this won't be the last time we need a fix in numpy.

Reply:
Do we want to use this `_IntegerType ` for all future type hints or `int` in SciPy? If yes we should move this to `_utils` or something else no? I am still confused about the need of this special type over `int` TBH.

Are you saying we should switch to NumPy nightlies here?

Reply:
> Do we want to use this `_IntegerType` for all future type hints or int in SciPy? If yes we should move this to `_utils` or something else no? 

We should move it to `numpy.typing` at some point.

> I am still confused about the need of this special type over int TBH.

It's because of the unfortunate existence of numpy array scalars - `np.int64(1)` is not a Python integer, but acts like one.

> Are you saying we should switch to NumPy nightlies here?

Better in a separate PR, unless it's necessary here because CI is failing.

Reply:
I am asking because we have also types in `differential_entropy` and we have used plain `int` there. Maybe we should communicate about new guidelines on how to type hint before we start adding "wrong" things.

Reply:
Unfortunatly, the nightly release is from 4 days ago (https://anaconda.org/scipy-wheels-nightly/numpy/files?sort=time&sort_order=desc). I am trying to build with @BvB93 modifications right now (so the main release).

Reply:
> Maybe we should communicate about new guidelines on how to type hint before we start adding "wrong" things.

We're going to find things like this as we go. How about we keep a wiki page with pointers for now? I feel like we're not at the point where it's easy/productive to write a section of the contributing guide yet.

Reply:
Just to say that with the news modifications, we have new `MyPy` errors unrelated to this PR:
```bash

scipy/special/orthogonal.pyi:273: error: Variable "numpy.poly1d" is not valid as a type  [valid-type]
scipy/special/orthogonal.pyi:273: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
scipy/special/orthogonal.pyi:273: error: Invalid base class "numpy.poly1d"  [misc]
scipy/fft/_pocketfft/tests/test_real_transforms.py:256: error: Invalid index type "Tuple[partial[Any], Type[int], int]" for "Dict[Tuple[partial[Any], Type[floating[Any]], int], int]"; expected type "Tuple[partial[Any], Type[floating[Any]], int]"  [index]
```

I must admit I feel a little bit lost. What should be the next steps ?

Reply:
Agreed @rgommers. I am not sure I can help here as I am struggling to get all these sorry. I know about normal typing but not NumPy specifics like these. I can review the document though if it can help.

Reply:
> Just to say that with the news modifications, we have new MyPy errors unrelated to this PR:
>
> ```
> scipy/special/orthogonal.pyi:273: error: Variable "numpy.poly1d" is not valid as a type  [valid-type]
> scipy/special/orthogonal.pyi:273: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
> scipy/special/orthogonal.pyi:273: error: Invalid base class "numpy.poly1d"  [misc]
> ```

Great, it appears that mypy does not allow the subclassing of objects annotated as `Type[Any]`, 
only explicitly defined classes. This is annoying, but it's something that we can fix in numpy with
a small followup on https://github.com/numpy/numpy/pull/18765/commits/b9d9e03014aaa4d7b6f06a89ceae06317a0012a3.